server/src/main/java/io/deephaven/server/hierarchicaltable/HierarchicalTableViewSubscription.java

```diff
@@ -365,11 +362,10 @@ private static void buildAndSendSnapshot(
             final boolean isFullSubscription = false;
             GrpcUtil.safelyOnNext(listener, streamGenerator.getSubView(
                     subscriptionOptions, initialSnapshot, isFullSubscription, rows, false,
-                    prevKeyspaceViewportRows, keyspaceViewportRows, columns));
-        }
+                    prevKeyspaceViewportRows, barrageMessage.rowsIncluded, columns));
 
-        prevKeyspaceViewportRows.clear();
-        prevKeyspaceViewportRows.insert(keyspaceViewportRows);
+            prevKeyspaceViewportRows.resetTo(barrageMessage.rowsIncluded);
```

Comment:
I think getSubView needs to document that (1) it does not own the row sets passed in, and (2) when they can be mutated.
The keyspace viewports can be mutated as soon as the call returns.
The position space viewport must not be mutated until the message is sent.